### PR TITLE
refactor: introduce `MonoType` in `Simplifier`

### LINKED ISSUE #6242: Introduce `MonoType` in `Simplifier`
... refactor all subsequent ASTs and phases.

The specific translation from `Type` to `MonoType` can be found in `Reducer` (and stolen/moved from there).

Instead of starting from SimplifiedAst, it would also be okay to "work backwards" from `Reducer`.

### PR DESCRIPTION
**Pull request by sockmaster27**
Fixes #6242

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 1]
```diff
@@ -45,30 +45,42 @@ object Simplifier {
...
-      case LoweredAst.Expr.Cst(cst, tpe, loc) => SimplifiedAst.Expr.Cst(cst, tpe, loc)
+      case LoweredAst.Expr.Cst(cst, tpe, loc) =>
+        val t = visitType(tpe)
+        SimplifiedAst.Expr.Cst(cst, t, loc)
 
       case LoweredAst.Expr.Lambda(fparam, exp, tpe, loc) =>
         val p = visitFormalParam(fparam)
         val e = visitExp(exp)
-        SimplifiedAst.Expr.Lambda(List(p), e, tpe, loc)
+        visitType(tpe) match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:72]: I understand the desire to do this, but we simply don't.

If we were to do it, we would do it consistently (e.g. everywhere for all expressions), but that turns out to be too much work.

Instead we simply use Type everywhere (MonoType) and that has worked out fine.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:72]: So I request to drop these special cases for Lambda and just use MonoType everywhere in SimplifiedAst.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 2]
```diff
@@ -78,60 +90,74 @@ object Simplifier {
...
+            val t = visitType(tpe)
+            SimplifiedAst.Expr.ApplyAtomic(op, es1, t, purity, loc)
 
           case AtomicOp.Spawn =>
-            // Wrap the expression in a closure: () -> tpe \ eff
+            // Wrap the expression in a closure: () -> tpe \ ef
             val List(e1, e2) = es
-            val lambdaTyp = Type.mkArrowWithEffect(Type.Unit, eff, e1.tpe, loc)
-            val fp = SimplifiedAst.FormalParam(Symbol.freshVarSym("_spawn", BoundBy.FormalParam, loc), Ast.Modifiers.Empty, Type.mkUnit(loc), loc)
+            val lambdaTyp = MonoType.Arrow(Nil, e1.tpe)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:111]: I think here `Nil` should be `List(Unit)`.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 3]
```diff
@@ -78,60 +90,74 @@ object Simplifier {
...
-            SimplifiedAst.Expr.ApplyAtomic(AtomicOp.Spawn, List(lambdaExp, e2), tpe, Purity.Impure, loc)
+            val t = visitType(tpe)
+            SimplifiedAst.Expr.ApplyAtomic(AtomicOp.Spawn, List(lambdaExp, e2), t, Purity.Impure, loc)
 
           case AtomicOp.Lazy =>
             // Wrap the expression in a closure: () -> tpe \ Pure
             val e = es.head
-            val lambdaTyp = Type.mkArrowWithEffect(Type.Unit, Type.Pure, e.tpe, loc)
-            val fp = SimplifiedAst.FormalParam(Symbol.freshVarSym("_lazy", BoundBy.FormalParam, loc), Ast.Modifiers.Empty, Type.mkUnit(loc), loc)
+            val lambdaTyp = MonoType.Arrow(Nil, e.tpe)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:120]: I think here `Nil` should be `List(Unit)`.

**Review by magnus-madsen — CHANGES_REQUESTED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala (hunk 1 above)
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:72]: Then we just need to do it in a later phase

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala (hunk 2 above)
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:111]: Both pass the tests

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:72]: We do? Where do we require Type.Lambda? I think it should just be `Type`?

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:111]: Yes, but List(Unit) is more correct.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala (hunk 1 above)
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:72]: In `LambdaLift.scala` we need to access `tpe.result`. Would it make more sense to handle it there?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:72]: Keep it local

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**